### PR TITLE
Implement W3DWaypointBuffer

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/ai/aistates.cpp [hunk 1]
```diff
@@ -53,3 +53,12 @@ bool Out_Of_Weapon_Range_Object(State *state, void *user_data)
     return false;
 #endif
 }
+
+const Coord3D *AIStateMachine::Get_Goal_Path_Position(int i) const
+{
+    if (i < 0 || i > m_goalPath.size()) {
```
Comment (feliwir) [inline src/game/logic/ai/aistates.cpp]: if `i` is 0 and the size of `m_goalPath` is also 0 this check will pass and line 63 will cause a segfault

File: src/game/logic/ai/aistates.h [hunk 2]
```diff
@@ -32,6 +32,9 @@ class AIStateMachine : public StateMachine
     virtual Utf8String Get_Current_State_Name() override;
 #endif
 
+    int Get_Goal_Path_Count() const { return m_goalPath.size(); }
```
Comment (feliwir) [inline src/game/logic/ai/aistates.h:35]: Should be inline or not?

File: src/game/logic/object/object.cpp [hunk 3]
```diff
@@ -286,6 +287,29 @@ void Object::On_Collide(Object *other, const Coord3D *loc, const Coord3D *normal
     }
 }
 
+ExitInterface *Object::Get_Object_Exit_Interface() const
+{
+    ExitInterface *exit = nullptr;
+
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
```
Comment (feliwir) [inline src/game/logic/object/object.cpp:294]: I'd use a range based for loop here or `std::find`

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+                                                    points[count] = Vector3(goal->x, goal->y, goal->z);
+                                                    count++;
+                                                    m_renderObj->Set_Position(Vector3(goal->x, goal->y, goal->z));
+                                                    W3D::Render(*m_renderObj, r);
+                                                    has_points = true;
+                                                }
+                                            }
+
+                                            if (has_points) {
+                                                m_segLine->Set_Color(Vector3(0.94999999f, 0.5f, 0.0f));
```
Comment (feliwir) [inline src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: this should be 0.95f, which will be represented as 0.9499999f, due to floating point arithmetics

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+                            if (rally2 != nullptr) {
+                                if (add_point) {
+                                    const GeometryInfo &info = obj->Get_Geometry_Info();
+                                    const Coord3D *pos2 = obj->Get_Position();
+                                    Coord3D pos3;
+                                    pos3.x = rally.x - pos.x;
+                                    pos3.y = rally.y - pos.y;
+                                    pos3.z = 0.0f;
+                                    pos3.Normalize();
+                                    pos3.Scale(99999.898f);
```
Comment (feliwir) [inline src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: 100000.0f

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+                                            coords[3].y = pos2->y - f2 - f3;
+
+                                            Coord2D *f6 = nullptr;
+                                            Coord2D *f7 = nullptr;
+                                            Coord2D *f8 = nullptr;
+                                            Coord3D pos6;
+                                            pos6.z = 0.0f;
+                                            Coord3D pos7;
+                                            pos7.z = 0.0f;
+                                            float f9 = 99999.898f;
```
Comment (feliwir) [inline src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: 100000.0f

**Review by feliwir — CHANGES_REQUESTED** [review]

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp (hunk 4 above)
Comment (feliwir) [reply src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: image

See https://www.h-schmidt.net/FloatConverter/IEEE754.html for reference :)

File: src/game/logic/ai/aistates.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/logic/ai/aistates.cpp]: Fixed, should have been >= not >

File: src/game/logic/object/object.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:294]: I prefer not to use range-based for, makes it harder to compare to the original pseudo.

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: fixed.

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp (hunk 5 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: fixed.

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp (hunk 6 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: fixed.

File: src/game/logic/object/update/aiupdate.cpp [hunk 7]
```diff
@@ -86,3 +86,27 @@ void AIUpdateInterface::Reset_Next_Mood_Check_Time()
     Call_Method<void, AIUpdateInterface>(PICK_ADDRESS(0x005D6030, 0x007FBAB5), this);
 #endif
 }
+
+unsigned int AIUpdateInterface::Get_AI_State_Type() const
```
Comment (tomsons26) [inline src/game/logic/object/update/aiupdate.cpp]: public: enum  AIStateType thiscall AIUpdateInterface::getAIStateType(void)const
see AIStateMachine ctor to map the enum

File: src/game/logic/object/contain/opencontain.h [hunk 8]
```diff
@@ -96,3 +96,38 @@ class ContainModuleInterface
...
+
+class ExitInterface
+{
+public:
+    virtual bool Is_Exit_Busy() const = 0;
+    virtual ExitDoorType Reserve_Door_For_Exit() = 0;
+    virtual void Exit_Object_Via_Door(Object *obj, ExitDoorType type) = 0;
+    virtual void Exit_Object_Via_Budding(Object *obj, Object *obj2) = 0;
+    virtual void Unreserve_Door_For_Exit(ExitDoorType type) = 0;
+    virtual void Exit_Object_In_A_Hurry() {}
```
Comment (tomsons26) [inline src/game/logic/object/contain/opencontain.h]: takes Object *

File: src/game/logic/object/contain/opencontain.h [hunk 9]
```diff
@@ -96,3 +96,38 @@ class ContainModuleInterface
...
+enum ExitDoorType
+{
+    EXIT_DOOR_UNK,
+};
+
+class ExitInterface
+{
+public:
+    virtual bool Is_Exit_Busy() const = 0;
+    virtual ExitDoorType Reserve_Door_For_Exit() = 0;
```
Comment (tomsons26) [inline src/game/logic/object/contain/opencontain.h]: takes ThingTemplate const*,Object *

File: src/game/logic/object/contain/opencontain.h (hunk 8 above)
Comment (jonwil) [reply src/game/logic/object/contain/opencontain.h]: fixed.

File: src/game/logic/object/contain/opencontain.h (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/contain/opencontain.h]: fixed.

File: src/game/logic/ai/aistates.cpp [hunk 10]
```diff
@@ -53,3 +53,12 @@ bool Out_Of_Weapon_Range_Object(State *state, void *user_data)
...
 }
+
+const Coord3D *AIStateMachine::Get_Goal_Path_Position(int i) const
+{
+    if (i < 0 || i >= m_goalPath.size()) {
+        return nullptr;
+    }
+
+    return &m_goalPath[i];
+}
```
Comment (xezon) [inline src/game/logic/ai/aistates.cpp]: eol

File: src/game/logic/ai/aistates.cpp [hunk 11]
```diff
@@ -53,3 +53,12 @@ bool Out_Of_Weapon_Range_Object(State *state, void *user_data)
     return false;
 #endif
 }
+
+const Coord3D *AIStateMachine::Get_Goal_Path_Position(int i) const
+{
+    if (i < 0 || i >= m_goalPath.size()) {
```
Comment (xezon) [inline src/game/logic/ai/aistates.cpp]: `int` compared with `size_t`. i expect compiler warning here. use static_cast.

File: src/game/logic/object/contain/opencontain.h [hunk 12]
```diff
@@ -96,3 +96,38 @@ class ContainModuleInterface
...
+    }
+
+    virtual bool Get_Exit_Position(Coord3D &coord)
+    {
+        coord.x = 0.0f;
+        coord.y = 0.0f;
+        coord.z = 0.0f;
+        return false;
+    }
+};
```
Comment (xezon) [inline src/game/logic/object/contain/opencontain.h]: eol

File: src/game/logic/object/update/aiupdate.cpp [hunk 13]
```diff
@@ -86,3 +86,27 @@ void AIUpdateInterface::Reset_Next_Mood_Check_Time()
...
 }
+
+unsigned int AIUpdateInterface::Get_AI_State_Type() const
+{
+    return Get_State_Machine()->Get_Current_State_ID();
+}
+
+int AIUpdateInterface::Friend_Get_Waypoint_Goal_Path_Size() const
+{
+    if (Get_AI_State_Type() != 6) {
```
Comment (xezon) [inline src/game/logic/object/update/aiupdate.cpp]: Do we have a names constant for this 6 ?

File: src/platform/w3dengine/client/w3dwaypointbuffer.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+#ifdef GAME_DLL
+    W3DWaypointBuffer *Hook_Ctor() { return new (this) W3DWaypointBuffer; }
+    void Hook_Dtor() { W3DWaypointBuffer::~W3DWaypointBuffer(); }
+#endif
+
+private:
+    RenderObjClass *m_renderObj;
+    SegmentedLineClass *m_segLine;
+    TextureClass *m_texture;
+};
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dwaypointbuffer.h]: eol

File: src/platform/w3dengine/client/w3dwaypointbuffer.h [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+class W3DWaypointBuffer
+{
+public:
+    W3DWaypointBuffer();
+    ~W3DWaypointBuffer();
+
+    void Free_Waypoint_Buffers();
+    void Set_Default_Line_Style();
+    void Draw_Waypoints(RenderInfoClass &rinfo);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dwaypointbuffer.h:32]: `const RenderInfoClass &rinfo` ?

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+void W3DWaypointBuffer::Draw_Waypoints(RenderInfoClass &rinfo)
+{
+    if (g_theInGameUI != nullptr) {
+        Set_Default_Line_Style();
+
+        if (g_theInGameUI->Is_In_Waypoint_Mode()) {
+            LightEnvironmentClass lightenv;
+            lightenv.Reset(Vector3(0.0f, 0.0f, 0.0f), Vector3(1.0f, 1.0f, 1.0f));
+            lightenv.Pre_Render_Update(rinfo.m_camera.Get_Transform());
+            RenderInfoClass r(rinfo.m_camera);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: rcamera

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+                        m_segLine->Set_Points(count, points);
+                        m_segLine->Render(rinfo);
+                    }
+                }
+            }
+        } else {
+            LightEnvironmentClass lightenv;
+            lightenv.Reset(Vector3(0.0f, 0.0f, 0.0f), Vector3(1.0f, 1.0f, 1.0f));
+            lightenv.Pre_Render_Update(rinfo.m_camera.Get_Transform());
+            RenderInfoClass r(rinfo.m_camera);
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: rcamera

File: src/game/logic/object/update/aiupdate.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.cpp]: fixed.

File: src/game/logic/ai/aistates.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/logic/ai/aistates.cpp]: fixed.

File: src/game/logic/ai/aistates.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/ai/aistates.cpp]: fixed.

File: src/game/logic/object/contain/opencontain.h (hunk 12 above)
Comment (jonwil) [reply src/game/logic/object/contain/opencontain.h]: fixed.

File: src/game/logic/object/update/aiupdate.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.cpp]: fixed.

File: src/platform/w3dengine/client/w3dwaypointbuffer.h (hunk 14 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dwaypointbuffer.h]: fixed.

File: src/platform/w3dengine/client/w3dwaypointbuffer.h (hunk 15 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dwaypointbuffer.h:32]: nope, it's not const.

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp (hunk 16 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: fixed.

File: src/platform/w3dengine/client/w3dwaypointbuffer.cpp (hunk 17 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dwaypointbuffer.cpp]: fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**